components/tor/tor_launcher_factory.cc

```diff
@@ -66,10 +66,10 @@ TorLauncherFactory::TorLauncherFactory()
     : is_starting_(false),
       is_connected_(false),
       tor_pid_(-1),
-      control_(new tor::TorControl(this->AsWeakPtr(),
-                                   content::GetIOThreadTaskRunner({})),
-               base::OnTaskRunnerDeleter(content::GetIOThreadTaskRunner({}))),
-      weak_ptr_factory_(this) {
+      control_(nullptr,
+               base::OnTaskRunnerDeleter(content::GetIOThreadTaskRunner({}))) {
+  control_.reset(new tor::TorControl(this->AsWeakPtr(),
```

Comment:
why does it have to reset? can't it just be in constructor initializer list?

Reply:
It can't. We can't get a weakptr before the object is fully constructed anymore. `this->AsWeakPtr()` would access an object that has not been constructed yet.